### ISSUE TITLE
Faulty behavior of "dayfirst" or another issue leading to wrong parsing

### ISSUE THREAD

**Issue by MathiasWedeken** [dateutil/dateutil#1334]
Title is unclear, since I don't know WHAT causes the issue. But "dayfirst=True" implies logically that month comes last in the string. But check this: 

```python
>>> parse("2020-1-10", dayfirst=True)
datetime.datetime(2020, 10, 1, 0, 0) #correct

>>> parse("2020-1-10", dayfirst=False)
datetime.datetime(2020, 1, 10, 0, 0) #correct

>>> parse("2020-09-30", dayfirst=True)
datetime.datetime(2020, 9, 30, 0, 0) #false, should give an error because month can't be bigger than 12

>>> parse("2020-09-30", dayfirst=False)
datetime.datetime(2020, 9, 30, 0, 0) #correct

parse("2020-30-09", dayfirst=True)
datetime.datetime(2020, 9, 30, 0, 0) #correct 

>>> parse("2020-30-09", dayfirst=False)
dateutil.parser._parser.ParserError: month must be in 1..12: 2020-30-09 #That's the expected output 
``` 
So, it looks like "something is wrong" when dayfirst=True. 
It could have been the range check that results in a parserError month must be in 1..12:  does not happen, but: 

```python 
>>> parse("2020-09-31", dayfirst=True)
dateutil.parser._parser.ParserError: day is out of range for month: 2020-09-31 #huh? what about dayfirst???  
```

**Comment by loafthecomputerphile** [comment]
Yeah i have the same issue and it seems for some reason that everyone is fine with it.

**Comment by Gorillatheif** [comment]
From parser._parse docstring:

::param dayfirst
Whether to interpret the first value in an ambiguous 3-integer date (e.g. 01/05/09) as the day (``True``) or month (``False``). 
If ``yearfirst`` is set to ``True``, this distinguishes between YDM and YMD. 

So seems like this parameter is not considered when providing a 4 character year, as in your examples.

**Comment by sphuber** [comment]
I don't think the position of the year really matters. Take the following example where the year is clearly at the end, it still gives counter-intuitive behavior:
```python
In [1]: date_parser.parse("13/07/2000", dayfirst=True)
Out[1]: datetime.datetime(2000, 7, 13, 0, 0) # correct

In [2]: date_parser.parse("13/07/2000", dayfirst=False)
Out[2]: datetime.datetime(2000, 7, 13, 0, 0) # incorrect, I would have expected this to raise

In [3]: date_parser.parse("12/07/2000", dayfirst=True)
Out[3]: datetime.datetime(2000, 7, 12, 0, 0) # correct

In [4]: date_parser.parse("12/07/2000", dayfirst=False)
Out[4]: datetime.datetime(2000, 12, 7, 0, 0) # correct
```
I would argue that the second example should raise and not silently ignore the `dayfirst` argument.

**Comment by sphuber** [comment]
Seems this is a duplicate of https://github.com/dateutil/dateutil/issues/214 . The solution requested there was to add an option to enforce strict behavior of `dayfirst` but it was never implemented.

**Comment by Gorillatheif** [comment]
> ```python
> In [2]: date_parser.parse("13/07/2000", dayfirst=False)
> Out[2]: datetime.datetime(2000, 7, 13, 0, 0) # incorrect, I would have expected this to raise
> ```
I agree. If I were debugging the above code in a vacuum, I'd be extremely confused.

I'm just trying to clarify the core of the confusion:
The ```dayfirst``` argument seems to have been designed to only be used as a last resort (when it is logically ambiguous if the day or month is first) rather than a prioritized rule.

Since "13/07/2000" has a 13 and a 7, it doesn't need to check ```dayfirst``` because the day is obviously the number greater than 12 and the month less than 12.

**Comment by sphuber** [comment]
Appreciate the clarification @Gorillatheif . It definitely makes sense if a human was parsing these date strings manually one by one. However, I think for many the use case of the `dateutil` parser is to automatically parse datetimes from a series of values, that may or may not share a common format. Issue #214 demonstrates this use case in the context of `pandas`. It tries to automatically detect the date format of a column using `dateutil` but even if `dayfirst` is explicitly set, it is not guaranteed that all values are interpreted with the same format since `dateutil.parser.parse` will silently ignore `dayfirst` if it does not make sense with the given value.

So far, it seems there is not a solution yet. In the mentioned issue, the current maintainer seemed to be onboard with implementing an option to make the parsing strict, but as far as I can tell, this was never actually completed. The only solution is to actually start implementing a custom parser, but that would defeat the purpose of having `dateutil` as a robust date parsing utility for the Python community.
